In [11]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

/kaggle/input/Militancy_News_Dataset.csv


In [12]:
import numpy as np
import pandas as pd

In [13]:
df=pd.read_csv("../input/Militancy_News_Dataset.csv",index_col ="gname")
df.head()


Unnamed: 0 addnotes scite1 scite2 scite3
gname                                                    
Sikh Extremists       21669      NaN    NaN    NaN    NaN
Unknown               22033      NaN    NaN    NaN    NaN
Unknown               29602      NaN    NaN    NaN    NaN
Unknown               33128      NaN    NaN    NaN    NaN
Sikh Extremists       33298      NaN    NaN    NaN    NaN

In [14]:
df.columns

Index(['Unnamed: 0', 'addnotes', 'scite1', 'scite2', 'scite3'], dtype='object')

In [15]:
df.shape

(2192, 5)

In [16]:
df.index.unique()

Index(['Sikh Extremists', 'Unknown', 'Muslim Militants',
       'Babbar Khalsa International (BKI)', 'Muslim Separatists',
       'Bodo Militants', 'Jammu and Kashmir Liberation Front',
       'Kashmiri extremists', 'Pro-Pakistani extremists',
       'Hizbul Mujahideen (HM)', 'Mujahideen Kashmir', 'Muslim Janbaz Force',
       'Pakistani trained terrorists', 'Al-Umar Mujahideen',
       'Achwan-I-Mushbani', 'Ikhwan-ul-Muslimeen', 'Muslim Brotherhood',
       'Kashmiri Hezbollah', 'Al-Fatah', 'Lashkari-e-Adam (Army of Adam)',
       'Allah's Tigers', 'Khalistan Zindabad (Long Live Khalistan)',
       'Minority Unity Forum', 'Harkat ul Ansar', 'Al-Jehad',
       'Afghan Merceneries', 'Muslims', 'Jamaat-E-Islami (India/Pakistan)',
       'Al-Faran', 'Dukhta-ran-e-Millat', 'Gunmen', 'Separatists',
       'Ikhwan Jammu and Kashmir', 'Jamiat ul-Mujahedin (JuM)',
       'Muslim Rebels', 'Tehrik al-Mojahedin', 'Lashkar-e-Taiba (LeT)',
       'Harakat ul-Mujahidin (HuM)', 'Militants', 'Muslim e

In [25]:
df_new = df.loc[['Jaish-e-Mohammad (JeM)','Lashkar-e-Taiba (LeT)','Hizbul Mujahideen (HM)']]
df_new.tail()
y=df_new.index
y.shape
df_new.head()

Unnamed: 0 addnotes  \
gname                                         
Jaish-e-Mohammad (JeM)       70230      NaN   
Jaish-e-Mohammad (JeM)       70413      NaN   
Jaish-e-Mohammad (JeM)       70414      NaN   
Jaish-e-Mohammad (JeM)       72388      NaN   
Jaish-e-Mohammad (JeM)       72988      NaN   

                                                                   scite1  \
gname                                                                       
Jaish-e-Mohammad (JeM)  “More on Suicide Attack in Kashmir,” Daily Exc...   
Jaish-e-Mohammad (JeM)  "Three Cops Among 5 Killed in Militant Attacks...   
Jaish-e-Mohammad (JeM)  "Three Cops Among 5 Killed in Militant Attacks...   
Jaish-e-Mohammad (JeM)  "Nine injured in militant bomb blast in Kashmi...   
Jaish-e-Mohammad (JeM)  “India: Militants trigger series of bomb blast...   

                                                                   scite2  \
gname                                                                       
Jaish-e-Mohammad (JeM)  “Report: Suicide Bomber Belongs to Azhar’s Mil...   
Jaish-e-Mohammad (JeM)  "Kashmir gun battles leave 14 dead", Reuters, ...   
Jaish-e-Mohammad (JeM)  "Kashmir gun battles leave 14 dead", Reuters, ...   
Jaish-e-Mohammad (JeM)  "Two die, 18 injured in string of blasts in In...   
Jaish-e-Mohammad (JeM)                                                NaN   

                       scite3  
gname                          
Jaish-e-Mohammad (JeM)    NaN  
Jaish-e-Mohammad (JeM)    NaN  
Jaish-e-Mohammad (JeM)    NaN  
Jaish-e-Mohammad (JeM)    NaN  
Jaish-e-Mohammad (JeM)    NaN

In [29]:
from sklearn.feature_extraction.text import TfidfVectorizer
v = TfidfVectorizer(stop_words='english')
x = v.fit_transform((df_new['addnotes']+df_new['scite1']+df_new['scite2']+df_new['scite3']).values.astype('U'))
x.shape

(306, 1057)

In [30]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.75,random_state=42)
dummies=pd.get_dummies(y_train)
dummies
dummies2=pd.get_dummies(y_test)
dummies2.head()

Hizbul Mujahideen (HM)  Jaish-e-Mohammad (JeM)  Lashkar-e-Taiba (LeT)
0                       1                       0                      0
1                       0                       0                      1
2                       0                       0                      1
3                       1                       0                      0
4                       0                       0                      1

## Using Logistic Regression

In [31]:
for c in [0.01, 0.05, 0.25, 0.5, 1]:
    
    lr = LogisticRegression(C=c)
    lr.fit(x_train, y_train)
    print ("Accuracy for C=%s: %s" 
           % (c, accuracy_score(y_test, lr.predict(x_test))))

Accuracy for C=0.01: 0.45454545454545453
Accuracy for C=0.05: 0.5194805194805194
Accuracy for C=0.25: 0.5584415584415584
Accuracy for C=0.5: 0.5454545454545454
Accuracy for C=1: 0.5584415584415584


## When we are using each word as a single word and predicting through logistic regression highest accuracy is coming to be 55.84% for c=0.25 and c=1

## Support Vector Machine

In [65]:
from sklearn.feature_extraction.text import CountVectorizer
ngram_vectorizer = CountVectorizer(binary=True, ngram_range=(1, 5),stop_words='english')
ngram_vectorizer.fit((df_new['addnotes']+df_new['scite1']+df_new['scite2']+df_new['scite3']).values.astype('U'))
x = ngram_vectorizer.transform((df_new['addnotes']+df_new['scite1']+df_new['scite2']+df_new['scite3']).values.astype('U'))
df_new = df.loc[['Jaish-e-Mohammad (JeM)','Lashkar-e-Taiba (LeT)','Hizbul Mujahideen (HM)']]
df_new.tail()
y=df_new.index
y.shape

x_train, x_test, y_train, y_test = train_test_split(x,y, train_size = 0.75,random_state=42)

for c in [0.01, 0.05, 0.25, 0.5, 1]:
    
    lr = LogisticRegression(C=c)
    lr.fit(x_train, y_train)
    print ("Accuracy for C=%s: %s" 
           % (c, accuracy_score(y_test, lr.predict(x_test))))


Accuracy for C=0.01: 0.5454545454545454
Accuracy for C=0.05: 0.5714285714285714
Accuracy for C=0.25: 0.5714285714285714
Accuracy for C=0.5: 0.5714285714285714
Accuracy for C=1: 0.5714285714285714


/opt/conda/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


## When we are using n grams where n=5 and predicting through logistic regression highest accuracy is coming to be 57.14% for c=0.05,0.25,0.5,1.In one case accuracy came to 62% percent but random state was not used and it became useless.

In [97]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.svm import LinearSVC


stop_words = ['in', 'of', 'at', 'a', 'the']
ngram_vectorizer = CountVectorizer(binary=True, ngram_range=(1, 3),stop_words=stop_words)
ngram_vectorizer.fit((df_new['addnotes']+df_new['scite1']+df_new['scite2']+df_new['scite3']).values.astype('U'))
x = ngram_vectorizer.transform((df_new['addnotes']+df_new['scite1']+df_new['scite2']+df_new['scite3']).values.astype('U'))
df_new = df.loc[['Jaish-e-Mohammad (JeM)','Lashkar-e-Taiba (LeT)','Hizbul Mujahideen (HM)']]
df_new.tail()
y=df_new.index
y.shape
x_train, x_test, y_train, y_test = train_test_split(x,y, train_size = 0.75,random_state=42)


for c in [0.001, 0.005, 0.01, 0.05, 0.1]:
    
    svm = LinearSVC(C=c)
    svm.fit(x_train, y_train)
    print ("Accuracy for C=%s: %s" 
           % (c, accuracy_score(y_test, svm.predict(x_test))))

Accuracy for C=0.001: 0.5584415584415584
Accuracy for C=0.005: 0.5714285714285714
Accuracy for C=0.01: 0.5714285714285714
Accuracy for C=0.05: 0.5714285714285714
Accuracy for C=0.1: 0.5714285714285714


## When we are using n grams where n=5 and predicting through support vector machine highest accuracy is coming to be 57.14% for c=0.05,0.25,0.5,1.In one case accuracy came to 62% percent but random state was not used and it became useless.

## Using Random Forest Classifier

In [127]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

stop_words = ['in', 'of', 'at', 'a', 'the']
ngram_vectorizer = CountVectorizer(binary=True, ngram_range=(1, 3),stop_words=stop_words)
ngram_vectorizer.fit((df_new['addnotes']+df_new['scite1']+df_new['scite2']+df_new['scite3']).values.astype('U'))
x = ngram_vectorizer.transform((df_new['addnotes']+df_new['scite1']+df_new['scite2']+df_new['scite3']).values.astype('U'))
df_new = df.loc[['Jaish-e-Mohammad (JeM)','Lashkar-e-Taiba (LeT)','Hizbul Mujahideen (HM)']]
df_new.tail()
y=df_new.index
y.shape
x_train, x_test, y_train, y_test = train_test_split(x,y, train_size = 0.75,random_state=42)


x_train, x_test, y_train, y_test = train_test_split(x,y, train_size = 0.75,random_state=42)

forest_model = RandomForestClassifier(random_state=1)
forest_model.fit(x_train, y_train)
print ("Accuracy for model: %s" 
        % (accuracy_score(y_test, svm.predict(x_test))))

Accuracy for model: 0.5714285714285714


/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


## Random Forest Classification is giving 57.14% accuracy